File: implementations/Quickfind.py
from .UnionFindAbtract import Template
# abstract base class template
# run like python -m implementations.Quickfind if you want to run the main
## quick find implementation


class QuickFind(Template):
    def __init__(self, n: int):
        self.id = list(range(n)) 

    def find (self,p):
        return self.id[p] # constant time find
    
    def connected(self, p, q):
        return self.find(p) == self.find(q) #true if connected
    
    def union(self, p: int, q: int) -> None:
        pid = self.find(p)
        qid = self.find(q)

        for i in range(len(self.id)): # changes all of the elements from one group to the next group when union is called
            if self.id[i] == pid:
                self.id[i] = qid


if __name__ == "__main__":
    uf = QuickFind(7)
    uf.union(1, 2)
    uf.union(2, 5)
    print(uf.id)
    print(uf.connected(1, 2))
    print(uf.connected(1, 4))
    print("test")

File: implementations/QuickFind.java
package implementations;
public class QuickFind {
    private int[] id;

    public QuickFind(int n) {
        id = new int[n];
        for (int i = 0; i < n; i++) {
            id[i] = i;
        }
    }

    public int find(int p) {
        return id[p]; // in constant time 
    }

    public boolean connected(int p, int q) {
        return find(p) == find(q); //T/F statement
    }
    
    public void union(int p, int q) {
        int pid = find(p);
        int qid = find(q);

        for (int i = 0; i < id.length; i++) { 
            if (id[i] == pid) {
                id[i] = qid;
            }
        }
    }

    //quick test to see if it works
    public static void main(String[] args) {
        QuickFind uf = new QuickFind(50);
        uf.union(12, 15);
        uf.union(33, 45);
        System.out.println(uf.connected(12, 15)); //should be true
        System.out.println(uf.connected(13, 45)); //should be false
        
    }
}